### PR TITLE
GT4Py backend options in StencilConfig

### PR DESCRIPTION
**Pull request by eddie-c-davis**
## Purpose

This PR implements a configuration file (`gt4py_options.yml`) in the `fv3core` package to abstract `gt4py` backend options from the `fv3core` code. The `StencilConfig` object is modified to read the configuration file and updates the backend options for matching backends. The format of each backend option entry in the file is as follows:
```
<option_name>:
  backend: <backend_name_pattern>
  value: <backend_option_value>
```

The initially defined options are `format_source` (all backends), `device_sync` (GPU backends), `skip_passes` (GTC backends), and `verbose` (compiled backends).

## Code changes:

- Default skip passes (`graph_merge_horizontal_executions` at present) moved from `decorators` to config file.
- Comments added to `fv_subgridz.init` and `ytp_v._ytp_v` to describe per stencil skipped passes
- `verbose` added to list of backend options in `future_stencil`
- Explicit `format_source` and `device_sync` options removed from `global_config`
- Add `global_config.read_backend_options_file` to read the YAML file
- Add `StencilConfig._get_backend_opts` to check whether the options from the config file should be applied to the active backend
- Refactor `StencilConfig` to support configuration options
- `format_source` and `device_sync` remain as optional constructor arguments so existing unit tests will pass
- Add new tests to `test_stencil_wrapper`

## Requirements changes:

- The `gt4py_options.yml` file is added to the `fv3core` package
- The `package_data` key is added to `setup.py` that points to the `gt4py_options.yml` file
- A `MANIFEST.in` file is added with the location of the `gt4py_options.yml` file

## Checklist
Before submitting this PR, please make sure:

- [x] You have followed the coding standards guidelines established at [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--BD7zigBMAhMZAPkeNENeuU2UAg-IlsYffZgTwyKEylty7NhY).
- [x] Docstrings and type hints are added to new and updated routines, as appropriate
- [x] Unit tests are added or updated for non-stencil code changes

### CODE REVIEW

**Comment by eddie-c-davis** [comment]
launch jenkins

**Comment by eddie-c-davis** [comment]
launch jenkins

File: fv3core/utils/global_config.py [hunk 1]
```diff
@@ -99,25 +107,53 @@ def __eq__(self, other):
...
+        device_sync: Optional[bool] = None,
+        format_source: Optional[bool] = None,
+    ) -> Dict[str, Any]:
+        if not StencilConfig._all_backend_opts:
+            StencilConfig._all_backend_opts = read_backend_options_file()
+        all_backend_opts: Dict[str, Any] = StencilConfig._all_backend_opts
+
+        candidate_opts: Dict[str, Any] = {}
+        if "all" in all_backend_opts:
+            candidate_opts.update(all_backend_opts["all"])
```
Comment (mcgibbon) [inline fv3core/utils/global_config.py:122]:
```suggestion
        candidate_opts.update(all_backend_opts.get("all", {}))
```

File: fv3core/utils/global_config.py [hunk 2]
```diff
@@ -99,25 +107,53 @@ def __eq__(self, other):
...
+
+        if func is not None:
+            stencil_name = f"{func.__module__.split('.')[-1]}.{func.__name__}"
+            if stencil_name in all_backend_opts:
+                candidate_opts.update(all_backend_opts[stencil_name])
+
+        backend_opts: Dict[str, Any] = {}
+        for name, option in candidate_opts.items():
+            if "backend" not in option or re.match(option["backend"], get_backend()):
+                backend_opts[name] = option["value"]
```
Comment (mcgibbon) [inline fv3core/utils/global_config.py:132]: The way this is currently written, you can only set the target value for a single backend. In other words, `name` here is unique across entries. Is this sufficient, or is it worth restructuring the yaml file to use list entries allowing configuration of the same name on different backends?

I would also suggest making a named variable e.g. `using_option_backend = re.match(pattern=option.get("backend", ""), string=get_backend())` then `if using_option_backend:` to make it a little clearer what this if statement is triggering on. I had to read it a few times before figuring it out.

**Comment by eddie-c-davis** [comment]
launch jenkins

File: fv3core/utils/global_config.py [hunk 3]
```diff
@@ -66,20 +55,30 @@ def is_gtc_backend() -> bool:
...
-        device_sync: bool,
+        format_source: Optional[bool] = None,
+        device_sync: Optional[bool] = None,
     ):
         self.backend = backend
         self.rebuild = rebuild
         self.validate_args = validate_args
-        self.format_source = format_source
-        self.device_sync = device_sync
+        self.backend_opts = self._get_backend_opts(device_sync, format_source)
```
Comment (FlorianDeconinck) [inline fv3core/utils/global_config.py:81]: `verbose` option?

Comment (eddie-c-davis) [reply fv3core/utils/global_config.py:81]: The `verbose` option is controlled through the YAML file, `device_sync` and `format_source` are included here to maintain backward compatibility with the existing tests. I can happily add `verbose` here too if you like, however.

**Pull request merged**